### PR TITLE
Patching Chef-Guard to work with Gitlab 11.4.10

### PR DESCRIPTION
**Pull request by andyrepton**
Fixes chef guard to use the API to download cookbooks

From:
https://sbp.gitlab.schubergphilis.com/$GROUP/$PROJECT/repository/archive.tar.gz

To:
https://sbp.gitlab.schubergphilis.com/api/v4/projects/$GROUP%2F$PROJECT/repository/archive.tar.gz

### CODE REVIEW

File: git/gitlab.go [hunk 1]
```diff
@@ -230,7 +230,7 @@ func (g *GitLab) GetDiff(project, user, sha string) (string, error) {
 
 // GetArchiveLink implements the Git interface
 func (g *GitLab) GetArchiveLink(project, tag string) (*url.URL, error) {
-	ns := fmt.Sprintf("%s/%s", g.group, project)
+	ns := fmt.Sprintf("%s%2F%s", g.group, project)
```
Comment (svanharmelen) [inline git/gitlab.go]: You should not change this. If you need this to be URL encoded, it should be done by using `url.QueryEscape(ns)` where the encoded version is needed. But in this case I don't think you should change this at all (see other comment).

File: git/gitlab.go [hunk 2]
```diff
@@ -246,7 +246,7 @@ func (g *GitLab) GetArchiveLink(project, tag string) (*url.URL, error) {
 	}
 
 	u, err := url.Parse(
-		fmt.Sprintf("/%s/repository/archive.tar.gz?ref=%s&private_token=%s",
+		fmt.Sprintf("/api/v4/projects/%s/repository/archive.tar.gz?ref=%s&private_token=%s",
```
Comment (svanharmelen) [inline git/gitlab.go]: I guess the URL did change (still would be nice if you have a changelog entry of the doc from Gitlab that details the exact change), but not to the API endpoint. Instead the URL should be updated to: 

```go
fmt.Sprintf("/%s/-/archive/%s/%s-%s.tar.gz?private_token=%s",
  ns,
  tag,
  project,
  tag,
  g.token,
)
```

**Review by svanharmelen — CHANGES_REQUESTED** [review]
Think this should help fixing the issue/problem?

File: git/gitlab.go (hunk 1 above)
Comment (andyrepton) [reply git/gitlab.go]: Fixed!

File: git/gitlab.go (hunk 2 above)
Comment (andyrepton) [reply git/gitlab.go]: Fair point, I'll make the changes

Comment (svanharmelen) [reply git/gitlab.go]: When you do, you should undo the `url.QueryEscape(ns)` change as that is not needed in this URL.

Comment (andyrepton) [reply git/gitlab.go]: I read the above too quickly. The issue here is that all GET requests that do not go to `/api/v4` without a valid cookie (even if you have an access token) are now returning 401 Unauthorized errors. Changing the URL to use the API is now the only way to download the archive file.

Comment (svanharmelen) [reply git/gitlab.go]: I see... In that case I suggest to make the URL relative instead of absolute and let the URL package take care of resolving the correct URL (using the call at the end of this package `g.client.BaseURL().ResolveReference(u)`):

```go
fmt.Sprintf("projects/%s/repository/archive.tar.gz?sha=%s&private_token=%s",
  url.QueryEscape(ns),
  tag,
  g.token,
)
```

Als note that `ref` needs to be changed to `sha` here ^^

**Comment by ghost** [comment]
How is this fix coming along?
I'm experiencing the same symptoms with our installation.
If there is any testing needed, let me know!

**Review by svanharmelen — APPROVED** [review]

**Comment by svanharmelen** [comment]
@bgroenendal are you able to compile the code in this PR yourself? As it would be nice to double check that the change works as expected now.

**Comment by ghost** [comment]
> @bgroenendal are you able to compile the code in this PR yourself? As it would be nice to double check that the change works as expected now.

I can try, not very familiar with go but I'll give it a 'go' (if that makes sense).

**Comment by ghost** [comment]
I'm probably doing something wrong.
Below the steps I'm using to compile the patch.

`# Install Go`
`https://golang.org/doc/install`

`# Clone master`
`cd ~/go/src/`
`git clone https://github.com/xanzy/chef-guard.git`

`# Apply patch`
`wget https://patch-diff.githubusercontent.com/raw/xanzy/chef-guard/pull/130.patch -O 130.patch`
`cd ~/go/src/chef-guard`
`git apply --stat ../130.patch`
`git apply --check ../130.patch`
`git am --signoff < ../130.patch`

`# Compile source`
`/usr/local/go/bin/go get`
`/usr/local/go/bin/go build`

`# Copy new binary`
`scp chef-guard chef.server:.`

This results in a chef-guard binary.
From there I replace original chef-guard binary and try to start the service. (so far so good)
Except that the chef-guard service won't start, it gives me the following error and goes in a failed state:

`chef-guard.service: main process exited, code=exit...XEC`

Am I missing some flags or steps to create a proper build?

**Comment by svanharmelen** [comment]
Let me see if I can help out by just writing out the steps...

```
Make sure Go is working as expected by typing `go version`:

$ go version
go version go1.11.5 darwin/amd64

$ mkdir -p ~/go/src/github.com/xanzy
$ cd ~/go/src/github.com/xanzy
$ git clone https://github.com/Seth-Karlo/chef-guard
$ cd chef-guard
$ git checkout fix_gitlab
$ make dev
$ scp bin/x64/chef-guard chef-server:
```

Does this help?

**Comment by ghost** [comment]
I've just done a test using the above steps.

This results in a working chef-guard file.

The bad part is that it didn't solve my problems regarding the connection to gitlab (either I'm looking at a different issue or this didn't solve the stated gitlab url issues)

**Comment by ghost** [comment]
I've done some more testing and the error I'm getting is rather odd. Not sure if this patch is involved but here are my findings.

When uploading I get the following error in my chef-guard.log:

`ERROR:   2019/03/06 16:38:15 Failed to send git spam: The token configured for GitLab group bkwi_chef-guard is not valid!`

This error comes from the following function:

- https://github.com/xanzy/chef-guard/blob/af64887f178b1ad061d78bb5fea38259ec334b26/git.go#L72

So, I'm doubting my token. But that one seems to be just fine when double (triple...) checking the token and my configuration. I can manually use the token to access the API and get my projects:

`curl -k --request GET --header "PRIVATE-TOKEN: *****" https://gitlab.mydomain.com/api/v4/search?scope=projects\&search=role`

In addition, I can see API actions happening in my GitLab logs + an actual commit in my Chef-Guard project changing e.g. the uploaded timestamp in the json for a cookbook. Or, even newly bootstrapped servers are in my activity stream.

I then started to tweak the following settings, just to see where this is coming from:

```
  commitchanges = true
  mailchanges = true
  searchgit = true
  publishcookbook = false
```
When I disable all, i get zero errors.
When I enable commitchanges on it's own, I get zero errors.
When enabling searchgit, I get zero errors.
The odd part is that I get the above when I enable mailchange.

So basically, in my setup it seems that the sending e-mails feature is causing problems.
These are my mail settings:

```
  maildomain = mydomain.com
  mailserver = 127.0.0.1
  mailport = 25
  mailrecipient = mailinglist@mydomain.com
  mailsendby =
```

Changing them won't change a thing.

So, long story short. I can see commits being made towards GitLab.
But sending e-mails seems to be broken for some reason (though it might still be something else).

Is there a way for me to go into debug mode for chef-guard to get the details behind the error that I'm getting?

**Comment by svanharmelen** [comment]
Does the owner of the token have access to the project that is queried for the diff? As it seems the error comes from this line: https://github.com/xanzy/chef-guard/blob/master/git/gitlab.go#L209

So the token is valid, but the token does not have permissions to get a diff for that project.

**Comment by ghost** [comment]
I've now tested with a token from the admin user, but even then I get that specific error.

This however isn't an issue with this patch, I have this behavior on both this patch and the current 0.7.2 release.

I'll do some more testing and will make a separate issue if needed.

**Comment by svanharmelen** [comment]
Thanks for the feedback @bgroenendal! Then I'll go ahead and merge this PR.

**Comment by ghost** [comment]
FYI. I found my issue as well. The gitconfig project was marked as private, changing this to public fixed my issue (which is fine in our case).

**Comment by svanharmelen** [comment]
Nice, glad to hear you found the problem 👍
